### ISSUE TITLE
Track LIBC file descriptors

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/libcx#43]
Currently, LIBCx tracks open files by file name using the using `FileDesc` struct when implementing additional processing (mmap, fcntl, pwrite etc) and doesn't keep track of which LIBC file descriptor numbers match these file names. This imposes a general problem of stale `FileDesc` structs kept around by LIBCx even when there are no open file descriptors and hence no potential operations that would actually need them, which in turn leads to quick exhausting of LIBCx shared memory in some usage scenarios (#30 is one of them).

A natural way to detect when a `FileDesc` becomes unused is to intercept the LIBC `close()` call. But given that there may be several file descriptors opened for the same file name and closing one of them doesn't invalidate all the others, it's impossible to simply free `FileDesc` or any part of it in an arbitrary `close()` call for the given file name — we can only do it if we are sure there are no other open file descriptors referring to the same file name. In order to implement that, we need to remember file descriptors that initiated creation or further access to a given `FileDesc` and use this information when processing the intercepted `close()` request.

Note that once we implement that, we will dramatically decrease memory usage so #30 should go away. Also, #7 won't be necessary to solve separately and #9 may be postponed further (because 2MB should be enough for most practical needs).

**Comment by dmik** [comment]
Note that currently we already try to free resources needed for our `fcntl` file lock enhancement but we do it incorrectly: `fcntl_locking_close()` will react on any `close()` call, not necessarily the one that closes the file descriptor passed to `fcntl` earlier. Of course, it's not typical that the same app opens the same file more than once but not typical doesn't mean impossible. So we should address it too.

Also note that the only way to free resources at this moment is to terminate all processes using LIBCx which will effectively free all memory and all resources occupied by LIBCx but given that LIBCx is getting used more and more, including various daemons (samba server, vnc server, ndctl plugins etc)., it's usually not practically possible w/o significant interruption of the user's work (which is not acceptable of course).

**Comment by dmik** [comment]
The code tracking file descriptors usage is implemented, it works (i.e. LIBCx data related to files, including pwrite mutexes and `FileDesc` structs themselves, gets freed) upon `close()` for the respective files, i.e. as soon as these files are no longer used by any of LIBCx processes. This should decrease memory and other resource usage in Samba scenario 10 times or so.

However, given that this is rather a major change in internal LIBCx structs, I had to carefully run the full test suite and fix small glitches here and there in order for all tests to work again. This is mostly done and I will commit after a final round of changes.

**Comment by dmik** [comment]
It took a while to bug fix the new implementation to make all test work under heavy load (running the full test suite in parallel 50-100 times in a tight loop). The new implementation, however, shows great results by reducing the number of used file structs from 13000 to just 70 and memory from 1.5M to just 8K ((i.e. almost 200 times less) during the Samba stress test (`git add -i` in Mozilla dir from Mac over Samba).

However, the Samba test still doesn't end successfully. The failure is not related to memory/resourced usage of course, it's a crash during one of `fcntl(F_GETLK)` calls Samba makes, in `get_file_desc_ex`, in the `desc->g->path` expression in `strcmp` because `desc->g` appears to be NULL for some reason. This isn't revealed by the test suite so I have to figure out on how to debug and catch it.

**Comment by dmik** [comment]
Also forgot to mention that when heavily testing the new implementation I spotted and fixed a couple of bugs that existed in the old code but they didn't receive separate commits because the major change was already there so it was rather hard to separate them.

**Comment by dmik** [comment]
Here is the crash report I get, JFTR:

[4013_01.TRP.txt](https://github.com/bitwiseworks/libcx/files/1252714/4013_01.TRP.txt)

**Comment by dmik** [comment]
The crash is fixed in 3b392ff4e00f62fe435328b5db82b56537a8dc7d (a miserable two lines mix). LIBCx appears rock solid now when using Samba: `git add -i` was running for like ~15 min and ended up successfully, Samba remained fully operational afterwards. it operated twice as 13 000 files it could do earlier before the screw up and memory still remained low (9K max). I think it's a perfect result for the task in question. Closing this.
